src/v/datalake/data_writer_interface.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/schemaless_translator.h"
+#include "iceberg/datatypes.h"
+#include "iceberg/values.h"
+
+#include <cstddef>
+#include <memory>
+
+#pragma once
+
+namespace datalake {
+class data_writer_result {
+public:
+    data_writer_result() = default;
+    data_writer_result(const data_writer_result&) = default;
+    data_writer_result(data_writer_result&&) = delete;
+    data_writer_result& operator=(const data_writer_result&) = default;
+    data_writer_result& operator=(data_writer_result&&) = delete;
+    virtual ~data_writer_result() = default;
+
+    virtual size_t get_row_count() = 0;
+};
+
+class data_writer {
+public:
+    data_writer() = default;
+    data_writer(const data_writer&) = default;
+    data_writer(data_writer&&) = delete;
+    data_writer& operator=(const data_writer&) = default;
+    data_writer& operator=(data_writer&&) = delete;
+    virtual ~data_writer() = default;
+
+    // TODO: error type?
+    virtual bool add_data_struct(
+      iceberg::struct_value&& /* data */, int64_t /* approx_size */)
+      = 0;
+
+    virtual std::unique_ptr<data_writer_result> finish() = 0;
```

Comment:
may it's better then to delete both copy and move constructors on `data_writer_result` if we are going to deal with them as unique pointers